include/layers/FCLayer.hpp

```diff
@@ -0,0 +1,124 @@
+#pragma once
+#include <stdexcept>
+#include <vector>
+
+template <typename ValueType>
+std::vector<ValueType> mat_vec_mul(
+    const std::vector<std::vector<ValueType> >& mat,
+    const std::vector<ValueType>& vec) {
+  std::vector<ValueType> res(mat.size());
+  ValueType elem;
+  for (size_t i = 0; i < mat.size(); i++) {
+    elem = ValueType(0);
+    for (size_t j = 0; j < vec.size(); j++) {
+      elem += mat[i][j] * vec[j];
+    }
+    res[i] = elem;
+  }
+  return res;
+}
+
+template <typename ValueType>
+class Layer {
+ public:
+  virtual std::vector<ValueType> run(const std::vector<ValueType>& input) = 0;
+};
+
+template <typename ValueType>
+class FCLayer : public Layer<ValueType> {
+ public:
+  FCLayer() : weights(), bias() {
+    inputSize = 0;
+    outputSize = 0;
+  }
+  FCLayer(const std::vector<std::vector<ValueType> >& input_weights,
+          const std::vector<ValueType>& input_bias);
+  FCLayer& operator=(const FCLayer& sec);
+  void set_weight(size_t i, size_t j, const ValueType& value) {
+    if (i >= outputSize || j >= inputSize) {
+      throw std::out_of_range("Bad weight index for FCLayer");
+    }
+    weights[i][j] = value;
+  }
+  ValueType get_weight(size_t i, size_t j) const {
+    if (i >= outputSize || j >= inputSize) {
+      throw std::out_of_range("Bad weight index for FCLayer");
+    }
+    return weights[i][j];
+  }
+  void set_bias(size_t i, const ValueType& value) {
+    if (i >= outputSize) {
+      throw std::out_of_range("Bad bias index for FCLayer");
+    }
+    bias[i] = value;
+  }
+  ValueType get_bias(size_t i) const {
+    if (i >= outputSize) {
+      throw std::out_of_range("Bad bias index for FCLayer");
+    }
+    return bias[i];
+  }
+  size_t get_input_size() { return inputSize; }
+  size_t get_output_size() { return outputSize; }
+  std::vector<ValueType> run(const std::vector<ValueType>& input);
+
+ private:
+  size_t inputSize;
+  size_t outputSize;
+  std::vector<std::vector<ValueType> > weights;
+  std::vector<ValueType> bias;
+};
+
+// weights * inputValues + bias = outputValues
+
+// constructor for FCLayer
+template <typename ValueType>
+FCLayer<ValueType>::FCLayer(
+    const std::vector<std::vector<ValueType> >& input_weights,
+    const std::vector<ValueType>& input_bias)
+    : weights(input_weights), bias(input_bias) {
+  if (input_weights.size() == 0) {
+    throw std::invalid_argument("Empty weights for FCLayer");
+  }
+  inputSize = input_weights[0].size();
+  outputSize = input_bias.size();
+  if (inputSize == 0 || outputSize == 0) {
+    throw std::invalid_argument("Bad weights/bias size for FCLayer");
+  }
+  // make weights isize x osize, filling empty with 0s
+  for (size_t i = 0; i < weights.size(); i++) {
+    while (weights[i].size() < inputSize) {
+      weights[i].push_back(ValueType(0));
+    }
+  }
```

Comment:
It looks like there is an option to use resize() from STL

Reply:
Fixed